db/model/generator.js

```diff
@@ -475,22 +570,29 @@ module.exports = function generator(seq, dataTypes) {
    * currentCollector field and expects the caller to save later.
    */
   Generator.prototype.assignToCollector = function () {
-    const possibleCollectors = this.possibleCollectors;
-    let newColl;
-    if (this.isActive && possibleCollectors && possibleCollectors.length) {
-      possibleCollectors.sort((c1, c2) => c1.name > c2.name);
-      newColl = possibleCollectors.find((c) => c.isRunning() && c.isAlive());
-    }
-
-    logAssignment(this.name, this.currentCollector, newColl);
-
-    // We could use setCurrentCollector, but that would result in database saves
-    // that would be unnecessary and complicate our logic in the db hooks.
-    // Instead, we set the foreign key (collectorId) from collector model.
-    // We also mock the currentCollector on the instance to avoid needing
-    // a database reload.
-    this.collectorId = newColl ? newColl.id : null;
-    this.currentCollector = newColl || null;
+    return Promise.resolve()
+    .then(() => {
+      const possibleCollectors = this.possibleCollectors;
+      if (this.isActive && possibleCollectors && possibleCollectors.length) {
+        return getMostAvailableCollector(possibleCollectors)
+        .then((mostAvailCollector) => {
+          logAssignment(this.name, this.currentCollector, mostAvailCollector);
```

Comment:
Shouldn't we check if we got a collector here since the fn *can* return "empty" Promise.resolve() on line 495?